src/Network_Equivalences-Communication.thy

```diff
@@ -13,6 +13,32 @@ lemma adapted_after_distributor:
   shows "(A \<Rightarrow> Bs) \<guillemotleft> \<E> = A \<guillemotleft> \<E> \<Rightarrow> map (\<lambda>B. B \<guillemotleft> \<E>) Bs"
   sorry
 
+lemma transition_from_distributor:
+  assumes "A \<Rightarrow> Bs \<rightarrow>\<^sub>s\<lparr>\<alpha>\<rparr> Q"
+  obtains n and X
+    where
+      "\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X"
+    and
+      "Q = \<Prod>B \<leftarrow> Bs. B \<guillemotleft> suffix n \<triangleleft> X \<parallel> (A \<Rightarrow> Bs) \<guillemotleft> suffix n"
+proof -
+  obtain n and X
+    where "\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X" and "Q = post_receive n X (\<lambda>x. \<Prod>B \<leftarrow> Bs. B \<triangleleft> \<box> x) \<parallel> (A \<Rightarrow> Bs) \<guillemotleft> suffix n"
+    unfolding distributor_def
+    using assms
+    by (fastforce elim: transition_from_repeated_receive)
+  moreover have "post_receive n X (\<lambda>x. B \<triangleleft> \<box> x) = B \<guillemotleft> suffix n \<triangleleft> X" for B
+    unfolding post_receive_after_send and post_receive_def
+    by (transfer, simp)
+  then have "post_receive n X (\<lambda>x. \<Prod>B \<leftarrow> Bs. B \<triangleleft> \<box> x) = \<Prod>B \<leftarrow> Bs. B \<guillemotleft> suffix n \<triangleleft> X"
+    by
+      (
+        induction Bs,
+        unfold general_parallel.simps post_receive_after_parallel post_receive_after_stop
+      ) simp_all
```

Comment:
This part obviously suffers from us having no lemma `post_receive_after_general_parallel` (you essentially prove this lemma on the fly in the second step). I just added issue input-output-hk/thorn-calculus#62 to get this lemma into the Þ-calculus implementation. Once it is there, the above code snippet should be compressed. I propose to change it to the following:
```isabelle
  moreover have "post_receive n X (λx. ∏B ← Bs. B ◃ □ x) = ∏B ← Bs. B « suffix n ◃ X"
    unfolding post_receive_after_general_parallel and post_receive_def
    by transfer simp
```

Reply:
Fixed in 7135b17086da1930642a00847ade0deeebc9f159, please verify.

Reply:
Looks good as well. :slightly_smiling_face: